lib/web/apps.go

```diff
@@ -161,31 +163,41 @@ func (h *Handler) getAppDetails(w http.ResponseWriter, r *http.Request, p httpro
 		return nil, trace.Wrap(err, "unable to resolve FQDN: %v", req.FQDNHint)
 	}
 
+	fqdn := result.FQDN
+
+	proxyPublicAddr := utils.InferProxyPublicAddr(req.FQDNHint, h.proxyDNSNames())
+	alwaysUseProxyPublicAddr := result.App.GetAlwaysUseProxyPublicAddr()
+	if alwaysUseProxyPublicAddr {
+		appName := strings.TrimSuffix(req.FQDNHint, proxyPublicAddr)
+		fqdn = utils.AssembleAppFQDN(appName, proxyPublicAddr, clusterName, result.App)
+	}
+
 	resp := &GetAppDetailsResponse{
-		FQDN: result.FQDN,
+		FQDN: fqdn,
 	}
 
 	requiredAppNames := result.App.GetRequiredAppNames()
 
 	if !isRedirectFlow {
-		// TODO (avatus) this would be nice if the string in the RequiredApps spec was the fqdn of the required app
-		// so we could skip the resolution step all together but this would break existing configs.
-
-		// if clusterName is not supplied in the params, the initial app must have been fetched with fqdn hint only.
-		// We can use the clusterName of the initially resolved app
-		if clusterName == "" {
-			clusterName = result.ClusterName
-		}
-		for _, required := range requiredAppNames {
-			res, err := h.resolveApp(r.Context(), ctx, ResolveAppParams{ClusterName: clusterName, AppName: required})
-			if err != nil {
-				h.logger.ErrorContext(r.Context(), "Error getting app details for associated required app", "required_app", required, "app", result.App.GetName())
-				continue
+		for _, requiredAppName := range requiredAppNames {
+			// craft the fqdn manually so we do not have to resolve required apps
+			if alwaysUseProxyPublicAddr {
+				requiredAppFQDN := fmt.Sprintf("%s.%s", requiredAppName, proxyPublicAddr)
```

Comment:
Do we need to verify/sanitize value of `requiredAppName` here? Also feel like it would be safe to resolve the app just so we do not send them to fqdn that may not resolve, but I lack proper insight here.

Reply:
we dont need to verify, this comes from the app spec (and was the same as before, i only updated the param name from "required" to "requiredAppName" in this loop for clarity)

if we resolve it here, it still resolves during the redirect loop for required apps. tbh, if its going to fail, i think its better to fail on the frontend as that error is more opaque to the user (instead of just logging and skipping as it did now).

but i think its preference at this point.